ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,271 @@
+            nref = Node(*nref_node)
+            self._execute(
+                "UPDATE cache SET nref=? WHERE id=?;", (None, nref.id),
+            )
+        elif pref_node:
+            pref = Node(*pref_node)
+            self._execute(
+                "UPDATE cache SET pref=? WHERE id=?;", (None, pref.id),
+            )
+
+    def _insert_item(self, key: Union[str, bytes], value: Any) -> None:
+        # insert new item into map and cache
+        pickled_value = self._pickle(value)
+        if self.is_empty:
+            new_pref = None
+            new_nref = None
+            new_pkey = 0
+
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?, ?);",
+                (new_pkey, key, pickled_value, new_pref, new_nref),
+            )
+
+            self._execute("INSERT INTO map VALUES (?, ?);", (key, new_pkey))
+
+        else:
+            # evict least recently used key/value
+            if self.is_full:
+                self.__delitem__(self.tail.key)
+
+            # get old head
+            old_head_pkey, _, _, old_head_pref, _ = self.head
+
+            # get attributes for new head
+            (max_pkey,) = self._fetch_single_query(
+                "SELECT id FROM cache ORDER BY id DESC LIMIT 1;"
+            )
+            new_pkey = max_pkey + 1
+            new_pref = None
+            new_nref = old_head_pkey
+
+            # add new head to cache
+            self._execute(
+                "INSERT INTO cache VALUES (?, ?, ?, ?, ?);",
+                (new_pkey, key, pickled_value, new_pref, new_nref),
+            )
+
+            # add new head to map
+            self._execute(
+                "INSERT INTO map VALUES (?, ?);", (key, new_pkey),
+            )
+
+            # update old head in cache
+            self._execute(
+                "UPDATE cache SET pref=? WHERE id=?;", (new_pkey, old_head_pkey),
+            )
+
+    def _update_item(self, key: Union[str, bytes], value: Any) -> None:
+        pickled_value = self._pickle(value)
+        self._execute("UPDATE cache SET value=? WHERE key=?;", (pickled_value, key))
+
+    def _fetch_single_query(self, query: str, args: Tuple[Any, ...] = None) -> Any:
+        with self._conn:
+            if args:
+                cursor = self._conn.execute(query, args).fetchall()
+            else:
+                cursor = self._conn.execute(query).fetchall()
+        if len(cursor) > 1:
+            raise Exception(
+                f"Invalid db state. More than one result found for query: {query}."
+            )
+        if not cursor:
+            return None
+        return first(cursor)
+
+    def _execute(self, query: str, args: Tuple[Any, ...] = None) -> None:
+        with self._conn:
+            if args:
+                self._conn.execute(query, args)
+            else:
+                self._conn.execute(query)
+
+    def _pickle(self, value: Any) -> bytes:
+        return sqlite3.Binary(pickle.dumps(value, -1))
```

Comment:
Can we allow for custom serializers to avoid the need for `pickle` which tends to be slow.

For most values we want to store we can do the serialization faster and just as easy.  This probably means requiring a *serializer* to be passed in which could be as simple as:

```python
TKey = TypeVar("TKey")
TValue = TypeVar("TValue")

class LRUSQLDict(MutableMapping[TKey, TMapping]):
    def __init__(self,
                 key_encoder: Callable[[TKey], bytes],
                 key_decoder: Callable[[bytes], TKey],
                 value_encoder: Callable[[TValue], bytes],
                 value_decoder: Callable[[bytes], TValue],
                 ...) -> None:
        ...
```

It's a bit verbose, but they are easy functions to provide and avoiding pickle, especially for things persisted in a database that are long lived is going to save us some headache.